client/src/main/java/org/apache/uniffle/client/impl/ShuffleWriteClientImpl.java

```diff
@@ -246,6 +259,18 @@ public SendShuffleDataResult sendShuffleData(String appId, List<ShuffleBlockInfo
     // which is minimum number when there is at most *replicaWrite - replica* sending server failures.
     for (ShuffleBlockInfo sbi : shuffleBlockInfoList) {
       List<ShuffleServerInfo> allServers = sbi.getShuffleServerInfos();
+      if (replica > 1 && !shuffleServerBlacklist.isEmpty()) {
+        // We can't modify allServers directly, because allServers is shared and is not thread safe
+        List<ShuffleServerInfo> newServerList = new ArrayList<>(allServers);
+        for (int i = newServerList.size() - 2; i >= 0; i--) {
+          ShuffleServerInfo serverInfo = newServerList.get(i);
+          if (shuffleServerBlacklist.contains(serverInfo)) {
+            newServerList.remove(i);
+            newServerList.add(serverInfo);
+          }
+        }
+        sbi.setShuffleServerInfos(newServerList);
```

Comment:
It seems that we don't need to call `setShuffleServerInfos`, we just need to call the `genServerToBlocks` with newServerList.

Reply:
It is just for UT now. And There may be other requirements need to modify this list in the future, such as #341.

Reply:
For #341 , we can also decide which server to use here instead of calling method `setShuffleServerInfos`.

Reply:
This method isn't thread safe, too.